Bio/bgzf.py

```diff
@@ -247,19 +247,56 @@
 binary mode, and decode the appropriate fragments yourself.
 """
 
+from __future__ import annotations
+
+import typing as tp
+
 import struct
 import sys
 import zlib
 
 from builtins import open as _open
 
+if tp.TYPE_CHECKING:
+    import _typeshed
+    import io
+    import collections.abc as cx_abc
+    from typing_extensions import TypeAlias, TypeGuard
+
+_BufferT = tp.TypeVar("_BufferT", str, bytes)
+_Self = tp.TypeVar("_Self")
+# fmt: off
+# Writing and updating modes
+_ReadingBinaryMode: TypeAlias = tp.Literal["rb", "br"]
+_ReadingTextMode: TypeAlias = tp.Literal["r", "tr", "rt"]
+_WritingAppendingBinaryMode: TypeAlias = tp.Literal["wb", "bw", "ab", "ba", "wb+", "w+b", "+wb", "bw+", "b+w", "+bw", "ab+", "a+b", "+ab", "ba+", "b+a", "+ba"]
+_WritingAppendingTextMode: TypeAlias = tp.Literal["w", "wt", "tw", "a", "at", "ta", "w+", "+w", "wt+", "w+t", "+wt", "tw+", "t+w", "+tw", "a+", "+a", "at+", "a+t", "+at", "ta+", "t+a", "+ta"]
+_reading_binary_modes: tp.Final[tuple[_ReadingBinaryMode, ...]] = tp.get_args(_ReadingBinaryMode)
+_reading_text_modes: tp.Final[tuple[_ReadingTextMode, ...]] = tp.get_args(_ReadingTextMode)
+_writing_appending_binary_modes: tp.Final[tuple[_WritingAppendingBinaryMode, ...]] = tp.get_args(_WritingAppendingBinaryMode)
+_writing_appending_text_modes: tp.Final[tuple[_WritingAppendingTextMode, ...]] = tp.get_args(_WritingAppendingTextMode)
+# fmt: on
+
+
 _bgzf_magic = b"\x1f\x8b\x08\x04"
 _bgzf_header = b"\x1f\x8b\x08\x04\x00\x00\x00\x00\x00\xff\x06\x00\x42\x43\x02\x00"
 _bgzf_eof = b"\x1f\x8b\x08\x04\x00\x00\x00\x00\x00\xff\x06\x00BC\x02\x00\x1b\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00"
 _bytes_BC = b"BC"
 
 
-def open(filename, mode="rb"):
+# fmt: off
+@tp.overload
+def open(filename: _typeshed.StrPath, mode: _ReadingBinaryMode = "rb") -> BgzfReader[bytes]: ...
+@tp.overload
+def open(filename: _typeshed.StrPath, mode: _WritingAppendingBinaryMode) -> BgzfWriter[bytes]: ...
+@tp.overload
+def open(filename: _typeshed.StrPath, mode: _ReadingTextMode) -> BgzfReader[str]: ...
+@tp.overload
+def open(filename: _typeshed.StrPath, mode: _WritingAppendingTextMode) -> BgzfWriter[str]: ...
+# fmt: on
+def open(
+    filename: _typeshed.StrPath, mode: str = "rb"
+) -> BgzfReader[str] | BgzfWriter[str] | BgzfReader[bytes] | BgzfWriter[bytes]:
```

Comment:
The addition of `from __future__ import annotations` ([PEP 563](https://peps.python.org/pep-0563/)) at the top of this module implicitly converts all annotations in the module to strings. Essentially, this means that typing features from newer versions of Python can be used in older versions of Python, as long as those features aren't grammar changes (causing syntax errors).

The `|` syntax is introduced in Python 3.10 in [PEP 604](https://peps.python.org/pep-0604/). However, because of the stringification of the annotations, this will work in Python 3.8.